core/impl/src/main/java/org/eclipse/sensinact/core/twin/impl/SensinactProviderImpl.java

```diff
@@ -137,4 +140,40 @@ public SensinactEMFService getOrCreateService(String name, EClass serviceEClass)
         }
         return new SensinactServiceImpl(active, this, provider, name, serviceInstance.eClass(), nexus, promiseFactory);
     }
+
+    @Override
+    public Provider getEMFProvider() {
+        checkValid();
+        return EcoreUtil.copy(provider);
+    }
+
+    @Override
+    public <T extends Provider> T getEMFProvider(Class<T> returnType) {
+        checkValid();
+        Objects.requireNonNull(returnType, "No EMF type given");
+
+        if (!returnType.isAssignableFrom(provider.getClass())) {
+            throw new ClassCastException(
+                    "EMF provider %s can't be cast to %s".formatted(getName(), returnType.getName()));
+        }
+
+        return returnType.cast(EcoreUtil.copy(provider));
+    }
+
+    @Override
+    public SensinactResource getResource(String service, String resource) {
+        Objects.requireNonNull(service, "No service name provided");
```

Comment:
Missing call to `checkValid`

```suggestion
        checkValid();
        Objects.requireNonNull(service, "No service name provided");
```